### PR TITLE
[Audit][Implementation] hasDraft replacement

### PR DESCRIPTION
**Pull request by kacper-mikolajczak**
### Details

This PR is part of the Callstack performance audit. It aims to remove `hasDraft` property from `Report` model and re-use existing collection of `ONYXKEY.COLLECTION.REPORT_DRAFT_COMMENT` to serve the same functionality. It tries to remove overhead coming from calculating draft status when writing a message in the composer.

### Fixed Issues

$ https://github.com/Expensify/App/issues/37880
PROPOSAL: https://expensify.slack.com/archives/C05LX9D6E07/p1709036231433249

### Tests

1. Open report
2. Write a draft message
3. Go back to LHN
4. Check if draft `pencil` icon is correctly displayed
5. Check if report position in the list has changed properly
6. Open the same report
7. Check if draft is correctly displayed
8. Remove message
9. Go back to LHN
10. Check if draft icon disappeared
11. Check if report position changed to original one

### QA Steps

Same as tests

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/62747088/088c9465-280f-4aeb-ac30-d03c67f5de78

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/62747088/6a7c71da-37c3-4583-a600-c341bd1b4766

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/62747088/1aa21fe8-aed6-4b22-95ca-0fcf68e77e98

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/62747088/b6fc8ca2-3d35-494e-ad7b-075eeaef4d52

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/62747088/8ea8e3e7-e8fa-404e-a0a5-e21e5a5cd11a

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/62747088/1a6f4f48-d136-4a9d-bdc7-c05a0825ac79

</details>

### CODE REVIEW

**Comment by kacper-mikolajczak** [comment]
Due to coupling between `Reports` and `gORIDs` method, I was not able to successfully extract the `draftReports` from `Reports`. Main reason is lack of ability to dedupe members of one collection and another when merging them together without affecting performance. Below, picture is presented with general overview and issue I've encountered.

image

**Comment by kacper-mikolajczak** [comment]
> Does the message containing only spaces should be saved as a draft?

By comparing behaviour with Messenger app, the code here does not treat spaces-only messages as valid drafts, thus they are not saved. It can be changed anytime - just let me know what is the descision 👍

File: src/libs/DraftCommentUtils.ts [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    key: ONYXKEYS.COLLECTION.REPORT_DRAFT_COMMENT,
+    callback: (value, key) => {
+        if (!key) {
+            return;
+        }
+
+        const reportID = key.replace(ONYXKEYS.COLLECTION.REPORT_DRAFT_COMMENT, '');
+        draftCommentMap[reportID] = value;
+    },
+});
```
Comment (hurali97) [inline src/libs/DraftCommentUtils.ts:16]: Can we do something like below? If we don't have `waitForCollectionCallback: true`, Onyx will do a loop for each key present in `REPORT_DRAFT_COMMENT`, which over the time can produce performance issues.

```suggestion
Onyx.connect({
    key: ONYXKEYS.COLLECTION.REPORT_DRAFT_COMMENT,
    waitForCollectionCallback: true,
    callback: (draftReportComments) => {
        draftCommentMap = draftReportComments;
    },
});
```

Later you can use `draftCommentMap` like:

```ts
function getDraftComment(reportID: string): OnyxEntry<string> {
    return draftCommentMap[`${ONYXKEYS.COLLECTION.REPORT_DRAFT_COMMENT}${reportID}`];
}
```

Comment (kacper-mikolajczak) [reply src/libs/DraftCommentUtils.ts]: Great suggestion @hurali97, thanks a lot! Done ✅

**Comment by kacper-mikolajczak** [comment]
Based on tests conducted on web, the performance difference for setting a draft is not noticeable. 

However, now draft calculations are **debounced** until user stops typing, instead of firing every time new message is being typed. This should improve UX by quite a margin and in situation where user types many messages fast, the solution evades `hasDraft` overhead completely.

**Comment by kacper-mikolajczak** [comment]
#### Profiling results
Below are the results of benchmarking _Sending a message_ functionality. Here is a recording of the performed test:


https://github.com/Expensify/App/assets/62747088/fdbef9fb-5465-4e50-b380-d4ecb7b92f63


**Web**
On web the aforementioned hypothesis does not hold true. The performance does not improve but does not degrade either.

after

before

**Comment by adhorodyski** [comment]
Thanks @kacper-mikolajczak for posting this together with traces. imo it looks ok - no degradation on web & solid mobile improvement, exactly what we initially aimed for.

**Comment by kacper-mikolajczak** [comment]
@shubham1206agra Should be alright ✅

File: tests/unit/SidebarOrderTest.ts [hunk 2]
```diff
@@ -780,7 +784,12 @@ describe('Sidebar', () => {
                     })
 
                     // When a new report is added
-                    .then(() => Onyx.merge(`${ONYXKEYS.COLLECTION.REPORT}${report4.reportID}`, report4))
+                    .then(() =>
+                        Promise.all([
```
Comment (kacper-mikolajczak) [inline tests/unit/SidebarOrderTest.ts:788]: @shubham1206agra in order to sync the merges in Onyx, I've used `Promise.all` here. Let me know if that's the correct way 🙏

Comment (kacper-mikolajczak) [reply tests/unit/SidebarOrderTest.ts:788]: @shubham1206agra AFAIK, there is no `multiSet` alternative (except for lower-level `Onyx.update`) so in order to merge those two values 'at once' (they are done separately but eventual consistency is what we care about), they are wrapped with `Promise.all`. I am not sure, if other methods would have exact same outcome. 

tl;dr; It waits for all the values to be merged successfully so the test can operate on up-to-date values

**Comment by kacper-mikolajczak** [comment]
Hi @shubham1206agra, is PR approved by you in general?

**Comment by kacper-mikolajczak** [comment]
Could you please review the PR as soon as I resolve the conflicts?

**Comment by kacper-mikolajczak** [comment]
@shubham1206agra Fixed ✅

**Review by shubham1206agra — APPROVED** [review]

**Review by shubham1206agra — APPROVED** [review]

**Comment by AndrewGable** [comment]
Please fix conflicts and make sure tests are passing, thanks!

**Comment by kacper-mikolajczak** [comment]
Hi @AndrewGable 👋 

> Please fix conflicts

Conflicts resolved ✅ 

> make sure tests are passing

The failing performance test is expected to happen (please see explanation [here](https://github.com/Expensify/App/pull/37281#issuecomment-2019919758)). We can set it as a new baseline 👍

**Review by AndrewGable — APPROVED** [review]
